build/lib/src/builder/build_step_impl.dart

```diff
@@ -60,14 +61,31 @@ class BuildStepImpl implements BuildStep {
 
   final void Function(Iterable<AssetId>)? _reportUnusedAssets;
 
-  BuildStepImpl(this.inputId, Iterable<AssetId> expectedOutputs, this._reader,
-      this._writer, this._resolvers, this._resourceManager,
+  final Future<PackageConfig> Function() _resolvePackageConfig;
+  Future<Result<PackageConfig>>? _resolvedPackageConfig;
+
+  BuildStepImpl(
+      this.inputId,
+      Iterable<AssetId> expectedOutputs,
+      this._reader,
+      this._writer,
+      this._resolvers,
+      this._resourceManager,
+      this._resolvePackageConfig,
       {StageTracker? stageTracker,
       void Function(Iterable<AssetId>)? reportUnusedAssets})
       : allowedOutputs = UnmodifiableSetView(expectedOutputs.toSet()),
         _stageTracker = stageTracker ?? NoOpStageTracker.instance,
         _reportUnusedAssets = reportUnusedAssets;
 
+  @override
+  Future<PackageConfig> get packageConfig async {
+    final resolved = _resolvedPackageConfig ??=
+        Result.capture(Future.sync(_resolvePackageConfig));
```

Comment:
It would guard against synchronous exceptions thrown by the callback, but that's not relevant here since we're passing an `async` function in `runBuilder`. I'll just invoke it directly.